### PR TITLE
Implement Collaborative Experts Model

### PR DESCRIPTION
**Pull request by Naqu6**
- Implement `Video Encoder` Model
- Implement `Text Encoder` Model
- Implement Temporal Aggregation Model
- Updated Loss Function and associated tests to handle multiple modalities better
- Added Gated Units for videos
- Added a NetVLAD from [Tensorflow addons](https://github.com/tensorflow/addons/pull/1237/commits) with Batch Normalization and Ghost Clusters
- Added Unit Tests for all code we've written

### CODE REVIEW

File: videoretrieval/metrics/loss.py [hunk 1]
```diff
@@ -16,29 +16,83 @@
...
+def build_similaritiy_matrix(
+    video_embeddings,
+    missing_experts,
+    text_embeddings,
+    mixture_weights):
+    """Builds a similarity matrix between text_embeddings and video_embeddings.
+
+    Arguments:
+        video_embeddings: a list of video embedding tensors, where each element
+            of the list is of shape batch_size x embedding dimensionality.
```
Comment (arkanath) [inline videoretrieval/metrics/loss.py:30]: nit: mention that the list is of length num_experts

File: videoretrieval/metrics/loss.py [hunk 2]
```diff
@@ -16,29 +16,83 @@
...
+    missing_experts_weights = 1 - tf.cast(missing_experts, tf.float32)
+
+    # Expand dimensions to allow for multiplication.
+    missing_experts_weights = tf.expand_dims(missing_experts_weights, 0)
+    mixture_weights = tf.expand_dims(mixture_weights, 1)
+
+    weights = mixture_weights * missing_experts_weights
+    weights, _ = tf.linalg.normalize(weights, axis=-1, ord=1)
+
+    similarity_matrix = 0
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/metrics/loss.py:54]: Thanks for the feedback! :) I'll make sure to change that.

File: videoretrieval/metrics/loss.py [hunk 3]
```diff
@@ -16,29 +16,83 @@
...
+
+    for expert_index, (
+        expert_video_embeddings, expert_text_embeddings) in enumerate(
+            zip(video_embeddings, text_embeddings)):
+
+        similarities = tf.matmul(
+            expert_text_embeddings, expert_video_embeddings, transpose_b=True)
+        similarities = similarities * weights[:, :, expert_index]
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/metrics/loss.py:62]: Yes, this should be element-wise multiplication. The text encoder outputs mixture weights, which are used at inference time to predict the weighting of the different expert video embeddings. See equation 5 in section 3.5 of [this paper](https://arxiv.org/pdf/1804.02516.pdf).

File: videoretrieval/models/components/textencoder.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            final embedding.
+        gems: a list of gated embedding modules, one per expert.
+        dense_layers: a list of dense layers, one per expert.
+
+    """
+    def __init__(self,
+            num_of_experts,
+            num_netvlad_clusters=25,
+            ghost_clusters=1,
+            language_model_dimensionality=768,
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/models/components/textencoder.py:53]: I'll just remove these default values: I had set them arbitrarily.

File: videoretrieval/models/components/textencoder.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+    def call(self, input_):
+        """Executes a forward pass on the text encoder.
+
+        First, the text is aggregated using netvlad. These aggregated
+        text embeddings are inputted to each gated embedding module to generate 
+        the normalized embeddings. The aggregated text embeddings are also
+        inputted into a dense layer to generate the mixture weights.
+
+        Parameters:
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/models/components/textencoder.py:118]: 👍

File: videoretrieval/models/components/textencoder.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            input_: a batch of contextual embeddings.
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/models/components/textencoder.py:119]: 👍

File: videoretrieval/metrics/loss.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (arkanath) [reply videoretrieval/metrics/loss.py:62]: similarities and weights[:, :, expert_index] are shape (batch_size, batch_size), you are doing matmul here but should this be element-wise instead?

File: videoretrieval/models/components/videoencoder.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+    ExpertProjectionModulationLayer, GatedEmbeddingUnitReasoning
+
+
+class VideoEncoder(tf.keras.Model):
+    """Implementation of the video encoder.
+
+    This model takes in a list of features computed by pretrained expert models
+    and produces a fixed length, sharded embeedding. This embedding corresponds
+    to the embedding produced by a corresponding text encoder. This model should
+    be trained in concert with a video encoder.
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/models/components/videoencoder.py:29]: 👍

File: videoretrieval/models/components/videoencoder.py [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import tensorflow as tf
+from models.layers import TemporalAggregationLayer, \
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/models/components/videoencoder.py:19]: Sure! Thanks for the catch.

File: videoretrieval/models/components/videoencoder.py [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+
+    Attributes:
+        num_experts: the number of pretrained experts used.
+        expert_aggregated_size: the dimensionality experts features are
+            projected to.
+        encoded_expert_dimensionality: the dimensionality aggregated experts
+            embeddings are mapped to.
+        temporal_aggregation_layers: a list of temporal aggregation layers, one
+            per expert.
+        g_mlp: A standard feedforward deep neural network, with g_mlp_layers.
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/models/components/videoencoder.py:44]: Totally! Thanks :)

File: videoretrieval/metrics/loss.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (arkanath) [reply videoretrieval/metrics/loss.py:70]: nvm, my bad, i didn't know * does element-wise :)

File: videoretrieval/tests/loss_tests.py [hunk 10]
```diff
@@ -25,69 +25,92 @@ class TestBidirectionalMaxMarginRankingLoss(unittest.TestCase):
...
             [1.0, 1.0],
             [0.6, 0.5],
             [0.9, 0.8],
-        ])
+        ])]
 
-        mock_text_embeddings = tf.Variable([
+        mock_bad_text_embeddings = [tf.constant([
             [-1.0, 0.0],
             [0, 1.0],
```
Comment (ozdemirbahadir) [inline videoretrieval/tests/loss_tests.py:103]: s/0/0.0

File: videoretrieval/models/layers/temporalaggregation.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        self.netvlad_clusters = netvlad_clusters
+
+        if self.use_netvlad:
+            self.netvlad = NetVLAD(self.netvlad_clusters, ghost_clusters)
+
+        self.projection_layer = tf.keras.layers.Dense(self.output_dim,
+            kernel_initializer=kernel_initializer,
+            bias_initializer=bias_initializer)
+
+    def call(self, input_):
```
Comment (ozdemirbahadir) [inline videoretrieval/models/layers/temporalaggregation.py:72]: Any reason for the trailing underscore?

File: videoretrieval/models/layers/netvlad.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        activation = self.fc(frames)
+
+        activation = self.batch_norm(activation)
+        activation = self.softmax(activation)
+
+        # remove ghost clusters
+        activation = activation[:, :self.num_clusters]
+        activation = tf.reshape(activation, (-1, max_frames, self.num_clusters))
+
+        # Soft-count of number of frames assigned to each cluster.
```
Comment (ozdemirbahadir) [inline videoretrieval/models/layers/netvlad.py:112]: Any better variable naming possible here? At least better than just 'a' and 'b'?

If not, please briefly explain in comments what they correspond to.

Comment (Naqu6) [reply videoretrieval/models/layers/netvlad.py:112]: Hey Bahadir, this implementation is from the Tensorflow addons package. I didn't pick the "a" and "b" variable names.

The only things I did were add Batch Normalization and Ghost Clusters. You can see the original PR here: https://github.com/tensorflow/addons/pull/1237/files.

Thanks :)

File: videoretrieval/models/layers/gatedembeddingmodule.py [hunk 13]
```diff
@@ -37,23 +36,105 @@ class GatedEmbeddingModule(tf.keras.layers.Layer):
...
+        embedding, mask = inputs
+
+        activations = self.fully_connected(embedding)
+
+        activations = self.batch_norm_one(activations)
+        mask = self.batch_norm_two(mask)
+
+        output = embedding * tf.nn.sigmoid(activations + mask)
+
+        return tf.math.l2_normalize(output, axis=-1)
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/models/layers/gatedembeddingmodule.py:140]: Totally!

File: videoretrieval/tests/loss_tests.py (hunk 10 above)
Comment (Naqu6) [reply videoretrieval/tests/loss_tests.py:103]: Thanks for the catch!

File: videoretrieval/models/layers/temporalaggregation.py (hunk 11 above)
Comment (Naqu6) [reply videoretrieval/models/layers/temporalaggregation.py:72]: Typically if I named something `input` I'd add a trailing underscore, otherwise I'd overwrite the built in function `input`. However, in this case I just renamed the variable :)

File: videoretrieval/metrics/loss.py [hunk 14]
```diff
@@ -16,29 +16,91 @@
...
+import numpy as np
+
+def build_similaritiy_matrix(
+    video_embeddings,
+    missing_experts,
+    text_embeddings,
+    mixture_weights):
+    """Builds a similarity matrix between text_embeddings and video_embeddings.
+
+    Arguments:
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/metrics/loss.py:28]: 👍

[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/metrics/loss.py:28]: Thanks! I'll add that :)

File: videoretrieval/metrics/loss.py [hunk 15]
```diff
@@ -16,29 +16,91 @@
 """
 
 import tensorflow as tf
+import numpy as np
+
+def build_similaritiy_matrix(
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/metrics/loss.py]: :+1 Fixed :)

File: videoretrieval/models/layers/netvlad.py (hunk 12 above)
Comment (ozdemirbahadir) [reply videoretrieval/models/layers/netvlad.py:112]: Acknowledged.

File: videoretrieval/models/layers/temporalaggregation.py (hunk 11 above)
Comment (ozdemirbahadir) [reply videoretrieval/models/layers/temporalaggregation.py]: Acknowledged.

File: videoretrieval/models/layers/gatedembeddingmodule.py [hunk 16]
```diff
@@ -37,23 +36,105 @@ class GatedEmbeddingModule(tf.keras.layers.Layer):
...
+        Returns: a tensor, l2 noramlized on the last dimension, of shape batch
+            size x output size.
+        """
+        assert len(inputs) == 2
+
+        embedding, mask = inputs
+
+        activations = self.fully_connected(embedding)
+
+        activations = self.batch_norm_one(activations)
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/models/layers/gatedembeddingmodule.py:135]: 👍
